landlab/components/flow_accum/lossy_flow_accumulator.py

```diff
@@ -0,0 +1,464 @@
+    >>> hmg.at_node['surface_water__discharge'] # doctest: +NORMALIZE_WHITESPACE
+    array([ 7.,  0.,  0.,
+            0.,  7.,  1.,  0.,
+            0.,  1.,  6.,  1.,  0.,
+            0., 1.,  1.,  0.,
+            0.,  0.,  0.])
+
+    With loss looks like this:
+
+    >>> def mylossfunction2(Qw, nodeID, linkID, grid):
+    ...     return (1. - grid.at_node['mylossterm'][nodeID]) * Qw
+    >>> fa = LossyFlowAccumulator(hmg, 'topographic__elevation',
+    ...                           flow_director=FlowDirectorSteepest,
+    ...                           depression_finder=DepressionFinderAndRouter,
+    ...                           loss_function=mylossfunction2)
+    >>> fa.run_one_step()
+    >>> hmg.at_node['drainage_area'] # doctest: +NORMALIZE_WHITESPACE
+    array([ 7.,  0.,  0.,
+            0.,  7.,  1.,  0.,
+            0.,  1.,  6.,  1.,  0.,
+            0., 1.,  1.,  0.,
+            0.,  0.,  0.])
+    >>> hmg.at_node['surface_water__discharge'] # doctest: +NORMALIZE_WHITESPACE
+    array([ 6.,  0.,  0.,
+            0.,  6.,  1.,  0.,
+            0.,  1.,  5.,  1.,  0.,
+            0., 1.,  1.,  0.,
+            0.,  0.,  0.])
+    >>> np.allclose(hmg.at_node['surface_water__discharge_loss'],
+    ...             lossy*hmg.at_node['surface_water__discharge'])
+    True
+
+    (Loss is only happening from the node, 14, that we set it to happen at.)
+
+
+    Finally, note we can use the linkIDs to create flow-length-dependent
+    effects:
+
+    >>> from landlab.components import FlowDirectorMFD
+    >>> mg = RasterModelGrid((4, 6), (2, 1))
+    >>> mg.set_closed_boundaries_at_grid_edges(True, True, False, True)
+    >>> z = mg.add_field('node', 'topographic__elevation', 2.*mg.node_x)
+    >>> z[9] = 8.
+    >>> z[16] = 6.5  # force the first node sideways
+
+    >>> L = mg.add_zeros('node', 'spatialloss')
+    >>> mg.at_node['spatialloss'][9] = 1.
+    >>> mg.at_node['spatialloss'][13] = 1.
+    >>> def fancyloss(Qw, nodeID, linkID, grid):
+    ...     # now a true transmission loss:
+    ...     Lt = (1. - 1./grid.length_of_link[linkID]**2)
+    ...     Lsp = grid.at_node['spatialloss'][nodeID]
+    ...     return Qw * (1. - Lt) * (1. - Lsp)
+
+    >>> fa = LossyFlowAccumulator(mg, 'topographic__elevation',
+    ...                           flow_director=FlowDirectorMFD,
+    ...                           loss_function=fancyloss)
+    >>> fa.run_one_step()
+
+    >>> mg.at_node['drainage_area'].reshape(mg.shape)
+    array([[  0. ,   0. ,   0. ,   0. ,   0. ,   0. ],
+           [  5.6,   5.6,   3.6,   2. ,   2. ,   0. ],
+           [ 10.4,  10.4,   8.4,   6.4,   4. ,   0. ],
+           [  0. ,   0. ,   0. ,   0. ,   0. ,   0. ]])
+    >>> mg.at_node['surface_water__discharge'].reshape(mg.shape)
+    array([[ 0. ,  0. ,  0. ,  0. ,  0. ,  0. ],
+           [ 4. ,  4. ,  2. ,  2. ,  2. ,  0. ],
+           [ 0. ,  8.5,  6.5,  4.5,  2.5,  0. ],
+           [ 0. ,  0. ,  0. ,  0. ,  0. ,  0. ]])
+    """
```

Comment:
I think its much better now that you just have the Lossy documentation!

Reply:
Cool